core/include/kernel/panic.h

```diff
@@ -29,9 +29,50 @@
 #define KERNEL_PANIC_H
 
 #include <compiler.h>
+#include <kernel/thread.h>
+#include <trace.h>
 
-#define panic()	__panic(__FILE__, __LINE__, __func__)
+static inline void __panic_prepare(void)
+{
+	(void)thread_mask_exceptions(THREAD_EXCP_ALL);
+}
 
-void __panic(const char *file, int line, const char *func) __noreturn;
+static inline __noreturn void __panic_do(void)
+{
+	while (1)
+		;
+}
+
+#if (CFG_TEE_CORE_DEBUG == 0)
+
+#define panic() \
+	do { \
+		__panic_prepare(); \
+		EMSG_RAW("Panic"); \
+		__panic_do(); \
+	} while (0)
+
+#define panic_trace(...)	panic()
+
+#else
+
+#define panic() \
+	do { \
+		__panic_prepare(); \
+		EMSG_RAW("Panic at %s:%d <%s()>", \
+			 __func__, __LINE__, __func__) ; \
+		__panic_do();	\
+	} while (0)
+
+#define panic_trace(...) \
+	do { \
+		__panic_prepare(); \
```

Comment:
If in a `__panic_xxx()` C function, we loose the `__VA_ARGS__`  and handle va_list. EMSG() can't be used and we need to call `trace_printf()`.
I would rather have `panic()` relaying on `EMSG(),/`EMSG_RAW()`` whatever prefixing/level/.. is set internally in trace layer.

Hence the `panic()` macro shall end in EMSG() carrying `__VA_ARGS__` at maco level.
to catch the file/line/func, the `panic()` shall expand straight __FILE__ etc...